File: Trees/Easy/LC_700-Search_in_a_Binary_Search_Tree/LC_700-Search_in_a_Binary_Search_Tree.cpp
// #include <bits/stdc++.h>
#include <iostream>
#include <cassert>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    // Iterative approach
    // T: O(log n) | O(h), M: O(1)
    // Where n is num of nodes, and h is height of resonably balanced tree
    TreeNode *searchBST(TreeNode *root, int val) {
        while (root) {
            if (root->val < val) 
                root = root->right;
            else if (root->val > val)
                root = root->left;
            else
                return root;
        }
        return NULL;
    }

    /*
    // Recursive approach
    // T: O(log n) | O(h), M: O(n)
    // Where n is num of nodes, and h is height of resonably balanced tree
    TreeNode *searchBST(TreeNode *root, int val) {
        if (!root)
            return NULL;
        
        if (root->val < val) 
            return searchBST(root->right, val);
        else if (root->val > val) 
            return searchBST(root->left, val);
        return root;
    }
    */
};

int main() {
    TreeNode root, *attempt;
    // Ex 1
    root = TreeNode(4, new TreeNode(2, new TreeNode(1), new TreeNode(3)), new TreeNode(7));
    attempt = Solution().searchBST(&root, 2);
    assert(attempt->val == 2);
    // Ex 2
    root = TreeNode(4, new TreeNode(2, new TreeNode(1), new TreeNode(3)), new TreeNode(7));
    attempt = Solution().searchBST(&root, 5);
    assert(attempt == NULL);

    return 0;
}

File: Trees/Easy/LC_700-Search_in_a_Binary_Search_Tree/LC_700-Search_in_a_Binary_Search_Tree.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution: 
    # Iterative approach
    # T: O(log n) | O(h), M: O(1)
    # Where n is num of nodes, and h is height of resonably balanced tree
    def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        while root:
            if root.val < val:
                root = root.right
            elif root.val > val:
                root = root.left
            else:
                return root
        return None

    '''
    # Recursive approach
    # T: O(log n) | O(h), M: O(1)
    # Where n is num of nodes, and h is height of resonably balanced tree
    def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if not root:
            return None
        
        if root.val < val:
            return self.searchBST(root.right, val)
        elif root.val > val:
            return self.searchBST(root.left, val)
        return root
    '''

sol = Solution()

# Ex 1
root = TreeNode(4, left=TreeNode(2, left=TreeNode(1), right=TreeNode(3)), right=TreeNode(7))
answer = root.left
attempt = sol.searchBST(root, 2)
assert attempt == answer, f'Expected 2 at {answer}, but got {attempt.val} at {attempt}'

# Ex 2
root = TreeNode(4, left=TreeNode(2, left=TreeNode(1), right=TreeNode(3)), right=TreeNode(7))
attempt = sol.searchBST(root, 5)
assert attempt == None, f'Expected NULL, but got {attempt.val}'
